crates/core/src/transport/packet_data.rs

```diff
@@ -113,51 +128,137 @@ impl<const N: usize> PacketData<N> {
             return false;
         }
         let mut is_intro_packet = true;
         // TODO: how many bytes do we need to check to be sure that it's not the intro packet?
         // for now we randomly check 64 bytes (intro_packet is 1500 bytes long)
         for i in (0..64).map(|_| thread_rng().gen_range(0..self.size)) {
             // TODO: use a fast rng here?
-            if self.data[i] != other.data[i] {
+            if self.data[i] != actual_intro_packet.data[i] {
                 is_intro_packet = false;
                 break;
             }
         }
         is_intro_packet
     }
+    
+    pub(super) fn convert_to_sym_enc(self) -> PacketData<SymmetricEncrypted, N> {
+        PacketData {
+            data: self.data,
+            size: self.size,
+            data_type: PhantomData,
+        }
+    }
+    
+    pub(super) fn convert_to_asym_enc(self) -> PacketData<AssymetricEncrypted, N> {
+        PacketData {
+            data: self.data,
+            size: self.size,
+            data_type: PhantomData,
+        }
+    }
 }
 
-impl<const N: usize> From<[u8; N]> for PacketData<N> {
-    fn from(data: [u8; N]) -> Self {
-        _check_valid_size::<N>();
-        Self { data, size: N }
+impl<const N: usize> PacketData<PlainText, N> {
+    pub fn from_bytes_to_plain(data: [u8; N], size : usize) -> Self {
+        Self {
+            data,
+            size,
+            data_type: PhantomData,
+        }
+    }
+}
+
+impl<const N: usize> PacketData<AssymetricEncrypted, N> {
+    pub fn from_bytes_to_asym_enc(data: [u8; N], size : usize) -> Self {
+        Self {
+            data,
+            size,
+            data_type: PhantomData,
+        }
+    }
+}
+
+impl<const N: usize> PacketData<SymmetricEncrypted, N> {
+    pub fn from_bytes_to_sym_enc(data: [u8; N], size : usize) -> Self {
+        Self {
+            data,
+            size,
+            data_type: PhantomData,
+        }
     }
 }
 
-impl<'a> From<&'a [u8]> for PacketData<MAX_PACKET_SIZE> {
-    fn from(value: &'a [u8]) -> Self {
-        let mut data = [0; MAX_PACKET_SIZE];
-        data[..value.len()].copy_from_slice(value);
+impl<const N: usize> PacketData<UnknownEncrypted, N> {
+    pub fn from_bytes_to_unknown_enc(data: [u8; N], size: usize) -> Self {
         Self {
             data,
-            size: value.len(),
+            size,
+            data_type: PhantomData,
         }
     }
 }
 
-impl<const N: usize> From<PacketData<N>> for Arc<[u8]> {
-    fn from(packet_data: PacketData<N>) -> Arc<[u8]> {
-        packet_data.data.into()
+impl<const N: usize> PacketData<PlainText, N> {
+    pub fn from_arc_to_plain(arc: Arc<[u8]>) -> Self {
+        let mut data = [0; N];
+        data.copy_from_slice(&arc);
+        Self {
+            data,
+            size: arc.len(),
+            data_type: PhantomData,
+        }
+    }
+}
+
+impl<const N: usize> PacketData<AssymetricEncrypted, N> {
+    pub fn from_arc_to_asym_enc(arc: Arc<[u8]>) -> Self {
```

Comment:
not sure we need this explicit conversions anywhere... instead of adding this functions try to do the conversions wherever necessary and that will tell you what methods are needed